File: src_js/grndwork_api_client/interfaces.ts
import {IterableResponse} from './iterable_response';

export interface RefreshToken {
  subject: string;
  token: string;
}

export interface ClientOptions {
  request_timeout?: number;
  request_retries?: number;
  request_backoff?: number;
}

export type DataValue = number | string | boolean | null;
export type QCValue = number | string | boolean | null;

export interface DataRecord {
  timestamp: string;
  record_num: number;
  data: Record<string, DataValue>;
  qc_flags?: Record<string, QCValue>;
}

export interface QCRecord {
  timestamp: string;
  qc_flags: Record<string, QCValue>;
}

export interface DataFileHeaders {
  meta?: Record<string, string>;
  columns: Array<string>;
  units: Array<string>;
  processing?: Array<string>;
}

export interface DataFile {
  source: string;
  source_start_timestamp: string | null;
  source_end_timestamp: string | null;
  filename: string;
  is_stale: boolean;
  headers: DataFileHeaders;
  created_at: string;
  updated_at: string;
}

export interface DataFileWithRecordsArray extends DataFile {
  records?: Array<DataRecord>;
}

export interface DataFileWithRecords extends DataFile {
  records: IterableResponse<DataRecord>;
}

export interface ProjectManager {
  full_name: string;
  email: string;
}

export interface Station {
  client_uuid: string;
  client_full_name: string;
  client_short_name: string;
  site_uuid: string;
  site_full_name: string;
  station_uuid: string;
  station_full_name: string;
  description: string;
  model: string;
  type: string;
  status: string;
  project_manager: ProjectManager | null;
  maintenance_frequency: string;
  maintenance_log: string;
  location_region: string;
  latitude: number;
  longitude: number;
  altitude: number;
  timezone_offset: number;
  start_timestamp: string | null;
  end_timestamp: string | null;
  data_file_prefix: string;
  created_at: string;
  updated_at: string;
}

export interface StationDataFile {
  filename: string;
  is_stale: boolean;
  headers: DataFileHeaders;
  created_at: string;
  updated_at: string;
}

export interface StationWithDataFiles extends Station {
  data_files: Array<StationDataFile>;
}

export interface DataExport {
  key: string;
  filename: string;
  format: string;
  format_options: Record<string, any>;
  headers: DataFileHeaders;
  start_timestamp: string;
  end_timestamp: string;
  record_count: number;
  status: string;
  created_at: string;
  updated_at: string;
}

export interface ReportFile {
  key: string;
  type: string;
  filename: string;
  description: string;
  created_at: string;
  updated_at: string;
}

export interface Report {
  key: string;
  package_name: string;
  kind: string;
  station_uuid: string;
  start_date: string;
  end_date: string;
  status: string;
  has_pdf: boolean;
  published_at: string;
  created_at: string;
  updated_at: string;
  data_exports: Array<DataExport>;
  files: Array<ReportFile>;
}

export interface GetStationsQuery {
  client?: string;
  site?: string;
  station?: string;
  limit?: number;
  offset?: number;
}

export interface GetReportsQuery {
  client?: string;
  site?: string;
  station?: string;
  before?: string;
  after?: string;
  limit?: number;
  offset?: number;
}

export interface GetDataFilesQuery {
  client?: string;
  site?: string;
  gateway?: string;
  station?: string;
  filename?: string;
  limit?: number;
  offset?: number;
}

export interface GetDataQuery extends GetDataFilesQuery {
  records_limit?: number;
  records_before?: string;
  records_after?: string;
}

export interface GetDataRecordsQuery {
  filename: string;
  limit?: number;
  before?: string;
  after?: string;
}

export interface GetDataQCQuery {
  filename: string;
  limit?: number;
  before?: string;
  after?: string;
}

export interface PostDataRecord {
  timestamp: string;
  record_num: number;
  data: Record<string, DataValue>;
}

export interface PostDataFile {
  filename: string;
  headers?: DataFileHeaders;
  records?: Array<PostDataRecord>;
}

export interface PostDataPayload {
  source: string;
  files: Array<PostDataFile>;
  overwrite?: boolean;
}

File: src_py/grndwork_api_client/interfaces.py
from typing import Dict, List, Optional, TypedDict, Union


class RefreshToken(TypedDict):
    subject: str
    token: str


class ClientOptions(TypedDict, total=False):
    request_timeout: float
    request_retries: int
    request_backoff: float


DataValue = Union[float, int, str, bool, None]
QCValue = Union[float, int, str, bool, None]


class _DataRecordRequired(TypedDict):
    timestamp: str
    record_num: int
    data: Dict[str, DataValue]


class DataRecord(_DataRecordRequired, total=False):
    qc_flags: Dict[str, QCValue]


class QCRecord(TypedDict):
    timestamp: str
    qc_flags: Dict[str, QCValue]


class _DataFileHeadersRequired(TypedDict):
    columns: List[str]
    units: List[str]


class DataFileHeaders(_DataFileHeadersRequired, total=False):
    meta: Dict[str, str]
    processing: List[str]


class DataFile(TypedDict):
    source: str
    source_start_timestamp: str | None
    source_end_timestamp: str | None
    filename: str
    is_stale: bool
    headers: DataFileHeaders
    created_at: str
    updated_at: str


class DataFileWithRecords(DataFile, total=False):
    records: List[DataRecord]


class ProjectManager:
    full_name: str
    email: str


class Station(TypedDict):
    client_uuid: str
    client_full_name: str
    client_short_name: str
    site_uuid: str
    site_full_name: str
    station_uuid: str
    station_full_name: str
    description: str
    model: str
    type: str  # noqa: A003
    status: str
    project_manager: ProjectManager | None
    maintenance_frequency: str
    maintenance_log: str
    location_region: str
    latitude: float
    longitude: float
    altitude: float
    timezone_offset: int
    start_timestamp: str | None
    end_timestamp: str | None
    data_file_prefix: str
    created_at: str
    updated_at: str


class StationDataFile(TypedDict):
    filename: str
    is_stale: bool
    headers: DataFileHeaders
    created_at: str
    updated_at: str


class StationWithDataFiles(Station):
    data_files: List[StationDataFile]


class GetStationsQuery(TypedDict, total=False):
    client: Optional[str]
    site: Optional[str]
    station: Optional[str]
    limit: Optional[int]
    offset: Optional[int]


class GetDataQuery(TypedDict, total=False):
    client: Optional[str]
    site: Optional[str]
    gateway: Optional[str]
    station: Optional[str]
    filename: Optional[str]
    limit: Optional[int]
    offset: Optional[int]
    records_limit: Optional[int]
    records_before: Optional[str]
    records_after: Optional[str]


class PostDataRecord(TypedDict):
    timestamp: str
    record_num: int
    data: Dict[str, DataValue]


class _PostDataFileRequired(TypedDict):
    filename: str


class PostDataFile(_PostDataFileRequired, total=False):
    headers: Optional[DataFileHeaders]
    records: Optional[List[PostDataRecord]]


class _PostDataPayloadRequired(TypedDict):
    source: str
    files: List[PostDataFile]


class PostDataPayload(_PostDataPayloadRequired, total=False):
    overwrite: Optional[bool]
